.github/workflows/openwrt.yml

```diff
@@ -1,112 +1,103 @@
-          ./scripts/version.sh
-        working-directory: n2n
+      - name: Build
+        uses: openwrt/gh-action-sdk@v5
+        env:
+          ARCH: ${{ matrix.arch }}
+          FEEDNAME: n2n_ci
+          PACKAGES: openwrt
 
-      - name: Copy n2n package definition into openwrt
-        run: |
-          cp -r n2n/packages/openwrt openwrt/package/n2n
-        working-directory: ./
+      - name: Move created packages to project dir
+        run: cp bin/packages/${{ matrix.arch }}/n2n_ci/*.ipk . || true
 
-      - name: Cache openwrt source downloads
-        uses: actions/cache@v2
+      - name: Store packages
+        uses: actions/upload-artifact@v2
         with:
-          path: openwrt/dl
-          key: openwrt-dl-${{ steps.openwrt_ref.outputs.REF }}
-
-      - name: Setup openwrt config and environment
-        run: |
-          echo "CONFIG_TARGET_x86=y" >.config
-          echo "CONFIG_TARGET_x86_64=y" >>.config
-
-      - name: Add n2n package to openwrt config
-        run: |
-          echo "CONFIG_PACKAGE_n2n-edge=m" >>.config
-          echo "CONFIG_PACKAGE_n2n-supernode=m" >>.config
+          name: ${{ matrix.arch}}-${{ github.sha}}-packages
+          path: "*.ipk"
 
-      - name: Build a full config from our stub file
-        run: |
-          make defconfig
+      - name: Store logs
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ matrix.arch}}-${{ github.sha}}-logs
+          path: logs/
 
-      - name: Download openwrt sources
-        run: |
-          make download
+      - name: Remove logs
+        run: sudo rm -rf logs/ || true
 
-      - name: Build openwrt build environment
+      - name: Register QEMU
+        if: ${{ matrix.runtime_test }}
         run: |
-          make -j `nproc` tools/install toolchain/install
+          sudo docker run --rm --privileged aptman/qus -s -- -p
 
-      - name: Build n2n dependancies
+      - name: Build Docker container
+        if: ${{ matrix.runtime_test }}
         run: |
-          make -j `nproc` package/libs/libpcap/compile
-
-      - name: Build n2n openwrt packages
+          docker build -t test-container --build-arg ARCH .github/openwrt/
         env:
-          N2N_PKG_VERSION: ${{ steps.n2n_ref.outputs.REF }}
-        run: |
-          echo "Build for $N2N_PKG_VERSION"
-          export N2N_PKG_VERSION
-          make package/n2n/clean V=s
-          make package/n2n/prepare USE_SOURCE_DIR=$GITHUB_WORKSPACE/n2n V=s
-          make package/n2n/compile V=s
-
-# FIXME: add a way to run the test suite!
-#      - name: Run embedded tests
-#        run: make test
+          ARCH: ${{ matrix.arch }}
 
-      - name: Upload built artifacts
-        uses: actions/upload-artifact@v2
-        with:
-          name: built-ipkgs
-          path: openwrt/bin/packages/*/base/*.ipk
+      - name: Test via Docker container
+        if: ${{ matrix.runtime_test }}
+        run: |
+          docker run --rm -v $GITHUB_WORKSPACE:/ci test-container
```

Comment:
> Failing that, it might be simpler if there were two different files - the file you are rewriting is not broken could exist side-by-side with your replacement. This would allow a comparison in actual use to be done.

I'm not sure what you want. You may simply change `Diff view` to `Split`.

Reply:
Additionally, it may well be that the existing file can still provide a good demonstration on how to do development of this code in an opwnwrt environment with quick cycle-times and access to the development codebase (which would be another reason to keep it).

If you want to do some developments, the proper approach is to download and use the SDK provided by upstream. In this case building toolchain from source is time-wasting IMHO.

> I can see that this is where Docker is used, and adding that to a build often has lead to making a codebase more difficult to quickly develop, debug and fix.

Mentioned above, just used for runtime tests. This is crossing-compile, we cannot do that on host.